### DISCUSSION TITLE
Question: Is it possible to run pyaudio on Ubuntu 22.04 under WSL2 with Windows 11?

### DISCUSSION THREAD

**Discussion by EdSeymore** [microsoft/WSL#9624] [General]
Would like to access the microphone and speakers from python3 application. I tried various approaches without success. I also was unable to get gradio to function with python3 on ubuntu. Also, the speakers do not honor my request to default to HDMI speakers when a USB micropohne with an earphone port is present without the earphone attached.

**Comment by bulutthecat** [comment]
I am having the same issue, is anybody planning to fix this? I use WSL to test applications before I deploy them to edge compute platforms.

**Reply by butzhang** [reply]
I have the same issue

**Reply by butzhang** [reply]
Update: I later bought a M3 pro as a workaround

**Reply by ZFYL** [reply]
sigma move
btw i found a way here: https://www.linuxuprising.com/2021/03/how-to-get-sound-pulseaudio-to-work-on.html
but still a lacking solution as of 2024

**Comment by EdSeymore** [comment]
Please see https://github.com/microsoft/WSL/issues/9896 for additional information. 

I have it running successfully with many ALSA lib messages being displayed.

**Reply by butzhang** [reply]
I am at the same place as you and have successfully installed pyaudio. and getting a stream of error from ALSA lib

**Reply by bulutthecat** [reply]
> Please see #9896 for additional information.
> 
> I have it running successfully with many ALSA lib messages being displayed.

There is an issue with ALSA lib, so far I have found a couple alternatives, but I would have to rewrite my program to support them. #9896 is only one of the issues regarding microphone usage in WSL. unless anybody knows of any good alternative I think we will have to wait.

**Reply by EdSeymore** [reply]
We should report this issue as a error with the WSL2 implementation. Having both Windows and Ubuntu running on the same computer seems like a very good arrangement going forward. Microsoft should put forth the effort to insure success of this implementation. This should not be left as an exercie for the end user.

**Reply by bulutthecat** [reply]
> We should report this issue as a error with the WSL2 implementation. Having both Windows and Ubuntu running on the same computer seems like a very good arrangement going forward. Microsoft should put forth the effort to insure success of this implementation. This should not be left as an exercie for the end user.

I agree, I have had similar issues before with passing on USB devices to WSL2. It would be a better idea to approach this more from a VM standpoint (proper pass-through for USB devices and disk mounting, along with GUI application support) then an in house OS implementation. At least then we could work with Xwindow and GUI applications without needing to start a server (and install a desktop environment). I am going to switch to a Raspberry PI for testing until these issues get resolved. I would love to see Microsoft pushing their Hyper-V virtualisation and making a better interface for WSL2 through their existing Hyper-V platform (which should be able to support all of this anyways).

**Comment by Jeronymous** [comment]
At the time being, has anyone succeeded to have pyaudio working in WSL2?

**Reply by ZFYL** [reply]
not yet
but it's getting frustrating
maybe we can have some kind of an audio service that feeds the audio stream from windows to a wsl2 audio client, that is a microphone

**Comment by r0kk** [comment]
If it will be of help to anyone, following worked for me after a lot of debugging in Ubuntu with `sounddevice `python library:
•	_Windows_: **Make sure that under Privacy & security, Microphone**, permissions are set **ON** for Let desktop apps access your microphone
•	_WSL2 Ubuntu:_ `sudo apt install libasound2-plugins`

wls and os version: `WSL version: 2.3.26.0 Kernel version: 5.15.167.4-1 WSLg version: 1.0.65 MSRDC version: 1.2.5620 Direct3D version: 1.611.1-81528511 DXCore version: 10.0.26100.1-240331-1435.ge-release Windows version: 10.0.22631.4460`

**Comment by suvarnak** [comment]
If it helps anyone, I think the confusion is due to difference in steps needed for WSL 1 and WSL2, and also due to the required setup on Windows firewall to allow incoming requests to the PulseAudio server running on Windows.

I used following python code to debug if I can access audio devices from my WSL. I saved it in as file and ran it from WSL cmdline.
```
import pyaudio

def list_audio_devices():
    p = pyaudio.PyAudio()
    for i in range(p.get_device_count()):
        info = p.get_device_info_by_index(i)
        print(f"Device {i}: {info['name']}")

if __name__ == "__main__":
    list_audio_devices()
```
As per various old tutorials, we need to follow certain installations on Windows and also on WSL.
1, PulseAudio service on Windows (by downloading windows binaries ) to act as Audio server. 
2. WSL should install pulseaudio packages (pulseaudio pulseaudio-utils) and also know the IP and should setup env variable PULSE_SERVER and also be allowed to access it 

However, I found that  if you are having Windows 11 and WSL 2, **you need not install packages on WSL** (pulseaudio pulseaudio-utils packages). Also I found that PulseAudio zip on freedesktop.org and installation steps does not work on Windows 11.

Following steps worked for me- 
**1. Install PulseAudio on Windows from binaries (Full Installer) at https://pgaskin.net/pulseaudio-win32/.** 

The setup using the zip file given at https://www.freedesktop.org/wiki/Software/PulseAudio/ did not work. I could get the pyaudio service started but WSL was not able to access audio. So I cleaned up the manual installation of PulseAudio on Windows, and reinstalled PulseAudio as a service with Full Installer at https://www.freedesktop.org/wiki/Software/PulseAudio/. The Full installer also help to setup access with Windows Firewall.

2. **On WSL2, all you need to set is env variable,** either on cmdline or in .bashrc
```export PULSE_SERVER=tcp:192.168.68.63```
Replace IP address `192.168.68.63` with IP address IPv4 - you can get it with ```ipconfig``` command on windows PS

After this, run the test script above on WSL comdline and you should see ```Device 0: pulse``` as output.

**Comment by matthiasdotsh** [comment]
For me https://github.com/bastibe/SoundCard works fine in WSL2.

**Comment by batu** [comment]
Hey folks, I was able to get it pyaudio Mic input running on WSL2. I used GPTs help, as such, I don't deeply understand the individual steps. 

However, I am sharing the steps taken below in case it helps someone:


### WSL2/WSLg: Fix PyAudio mic error “[Errno -9996] Invalid input device (no default output device)” (Deepgram or any PyAudio app)

- **Problem**
  - PyAudio/PortAudio can’t see a default input in WSL2 → “[Errno -9996] Invalid input device (no default output device)”.
  - Cause: ALSA has no default device; WSLg exposes audio via PulseAudio (RDP), but ALSA isn’t routed to Pulse by default.

### Steps

1) Verify you’re on WSLg (Pulse socket exists)
```bash
test -S /mnt/wslg/PulseServer && echo "WSLg OK" || echo "No WSLg"
```
- Learn more: WSLg audio architecture [docs](https://github.com/microsoft/wslg/blob/main/docs/WSLg-Architecture.md)

2) Install audio client libs
```bash
sudo apt-get update
sudo DEBIAN_FRONTEND=noninteractive apt-get install -y \
  libportaudio2 libasound2t64 libasound2-plugins alsa-utils pulseaudio-utils
```
- Learn more: Ubuntu sound troubleshooting [guide](https://help.ubuntu.com/community/SoundTroubleshooting) and PulseAudio [wiki](https://wiki.ubuntu.com/PulseAudio)

3) Route ALSA → PulseAudio (create `~/.asoundrc`)
```bash
cat > ~/.asoundrc <<'EOF'
pcm.!default { type pulse fallback "sysdefault" hint.description "Default Audio Device (via PulseAudio)" }
ctl.!default { type pulse fallback "sysdefault" }
EOF
```
- Learn more: PulseAudio ALSA plugin examples (ArchWiki) [link](https://wiki.archlinux.org/title/PulseAudio/Examples#ALSA)

4) Point apps to WSLg’s Pulse server
```bash
echo 'export PULSE_SERVER=unix:/mnt/wslg/PulseServer' >> ~/.bashrc
source ~/.bashrc
```
- Learn more: `PULSE_SERVER` (manpage) [link](https://manpages.ubuntu.com/manpages/noble/en/man5/pulse-client.conf.5.html)

5) Sanity check Pulse and recording
```bash
pactl info | sed -n '1,15p'         # should show Default Source: RDPSource, Default Sink: RDPSink
pactl list sources short
arecord -D default -f S16_LE -c 1 -r 16000 -d 2 /tmp/test.wav && ls -lh /tmp/test.wav
```
- Learn more: `pactl` [manpage](https://www.freedesktop.org/software/pulseaudio/man/pactl.html), `arecord` [manpage](https://manpages.ubuntu.com/manpages/noble/en/man1/arecord.1.html)

6) Run your PyAudio app (Deepgram example)
```bash
uv run deepgram/live-streaming-starter-kit/test_suite.py -k <YOUR_DEEPGRAM_KEY> -i mic
```
- Learn more: Deepgram real-time streaming [docs](https://developers.deepgram.com/docs/streaming), PyAudio [docs](https://people.csail.mit.edu/hubert/pyaudio/)

7) Optional: hide ALSA warnings (they’re harmless)
```bash
uv run deepgram/live-streaming-starter-kit/test_suite.py -k <YOUR_DEEPGRAM_KEY> -i mic 2>/dev/null
```
- Learn more: Bash redirections [manual](https://www.gnu.org/software/bash/manual/bash.html#Redirections)

### Why this works
- PyAudio uses PortAudio, which targets ALSA on Linux.
- In WSL2, there are no native ALSA cards; WSLg exposes audio via PulseAudio’s RDP devices (`RDPSource`, `RDPSink`).
- `~/.asoundrc` makes ALSA’s default device a Pulse shim, and `PULSE_SERVER` points to WSLg’s Pulse socket so PyAudio can open the mic.

### Quick troubleshooting
- Enable Windows mic permission for your terminal app (Windows Settings → Privacy & security → Microphone).
- Confirm `pactl list sources short` shows `RDPSource`.
- If `/mnt/wslg/PulseServer` is missing (no WSLg), run the Python app on Windows or set up a different Pulse route.

**Reply by GuillaumeRattin** [reply]
Thank you so much it worked!

**Reply by KanishkaVerma054** [reply]
Thanks, It worked so well!
